src/traceViewer.ts

```diff
@@ -17,67 +17,79 @@
 import { ChildProcess, spawn } from 'child_process';
 import type { TestConfig } from './playwrightTestTypes';
 import { SettingsModel } from './settingsModel';
-import { findNode } from './utils';
+import { findNode, getNonce } from './utils';
 import * as vscodeTypes from './vscodeTypes';
+import { DisposableBase } from './disposableBase';
 
-export class TraceViewer implements vscodeTypes.Disposable {
+function getPath(uriOrPath: string | vscodeTypes.Uri) {
+  return typeof uriOrPath === 'string' ?
+    uriOrPath :
+    uriOrPath.scheme === 'file' ?
+      uriOrPath.fsPath :
+      uriOrPath.toString();
+}
+
+export type TraceViewer = SpawnTraceViewer | EmbeddedTraceViewer;
+
+export class SpawnTraceViewer extends DisposableBase {
   private _vscode: vscodeTypes.VSCode;
   private _envProvider: () => NodeJS.ProcessEnv;
-  private _disposables: vscodeTypes.Disposable[] = [];
   private _traceViewerProcess: ChildProcess | undefined;
   private _settingsModel: SettingsModel;
-  private _currentFile?: string;
+  private _config: TestConfig;
+  private _serverUrlPrefix?: string;
 
-  constructor(vscode: vscodeTypes.VSCode, settingsModel: SettingsModel, envProvider: () => NodeJS.ProcessEnv) {
+  constructor(vscode: vscodeTypes.VSCode, settingsModel: SettingsModel, envProvider: () => NodeJS.ProcessEnv, config: TestConfig) {
+    super();
     this._vscode = vscode;
     this._envProvider = envProvider;
     this._settingsModel = settingsModel;
+    this._config = config;
 
     this._disposables.push(settingsModel.showTrace.onChange(value => {
       if (!value && this._traceViewerProcess)
         this.close().catch(() => {});
     }));
   }
 
-  currentFile() {
-    return this._currentFile;
+  isEnabled() {
+    return this._settingsModel.showTrace.get();
   }
 
-  async willRunTests(config: TestConfig) {
-    if (this._settingsModel.showTrace.get())
-      await this._startIfNeeded(config);
+  isStarted() {
+    return !!this._traceViewerProcess;
   }
 
-  async open(file: string, config: TestConfig) {
-    if (!this._settingsModel.showTrace.get())
-      return;
-    if (!this._checkVersion(config))
+  async willRunTests() {
```

Comment:
I'd try pulling this to the call site as well. Think about it this way: SpawnTraceViewer and EmbedTraceViewer know as little about the embedder as possible, including the settings. An external entity picks the viewer based on the setting, calls its willRunTests, etc.

Reply:
addressed in https://github.com/microsoft/playwright-vscode/pull/509